types/scene/subsurface_tree.c

```diff
@@ -0,0 +1,218 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_scene.h>
+#include <wlr/util/addon.h>
+
+/**
+ * A tree for a surface and all of its child sub-surfaces.
+ *
+ * `tree` contains `scene_surface` and one node per sub-surface.
+ */
+struct wlr_scene_subsurface_tree {
+	struct wlr_scene_tree *tree;
+	struct wlr_surface *surface;
+	struct wlr_scene_surface *scene_surface;
+
+	struct wlr_scene_subsurface_tree *parent; // NULL for the top-level surface
+	struct wlr_addon surface_addon; // only set if there's a parent
+
+	struct wl_listener tree_destroy;
+	struct wl_listener surface_destroy;
+	struct wl_listener surface_commit;
+	struct wl_listener surface_new_subsurface;
+};
+
+static void subsurface_tree_destroy(struct wlr_scene_subsurface_tree *subsurface_tree) {
+	// tree and scene_surface will be cleaned up by scene_node_finish
+	if (subsurface_tree->parent) {
+		wlr_addon_finish(&subsurface_tree->surface_addon);
+	}
+	wl_list_remove(&subsurface_tree->tree_destroy.link);
+	wl_list_remove(&subsurface_tree->surface_destroy.link);
+	wl_list_remove(&subsurface_tree->surface_commit.link);
+	wl_list_remove(&subsurface_tree->surface_new_subsurface.link);
+	free(subsurface_tree);
+}
+
+static void subsurface_tree_handle_tree_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_subsurface_tree *subsurface_tree =
+		wl_container_of(listener, subsurface_tree, tree_destroy);
+	subsurface_tree_destroy(subsurface_tree);
+}
+
+static void subsurface_tree_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_subsurface_tree *subsurface_tree =
+		wl_container_of(listener, subsurface_tree, surface_destroy);
+	wlr_scene_node_destroy(&subsurface_tree->tree->node);
+}
+
+static const struct wlr_addon_interface subsurface_tree_addon;
```

Comment:
Nit: every other use of `wlr_addon_interface` has `*_addon_impl` name pattern AFAIK

Reply:
Fixed!